torch/utils/data/datapipes/utils/snapshot.py

```diff
@@ -0,0 +1,48 @@
+from torch.utils.data.datapipes.datapipe import IterDataPipe
+from torch.utils.data.graph_settings import apply_shuffle_seed
+
+
+# TODO: Caveats
+#   1. Caller (either the ReadingService or DataLoader) must pass in the initial RNG
+#   2. `in_batch_shuffle` and `bucketbatch` are not compatible with this because they currently
+#      lack the option to `set_seed`.
+
+def _simple_snapshot_graph(datapipe: IterDataPipe, n_iterations: int, rng=None) -> None:
+    r"""
+    This function will restore a snapshot by fast-forwarding the given DataPipe by ``n_iterations``,
+    and in the process, fast-forward its parent DataPipes as well at the cost of re-doing every computation.
+    For instance, applying this function to the final DataPipe of a graph will restore the snapshot
+    (via fast-forward) every DataPipe within the graph.
+
+    Note:
+        This is the simplest but least efficient way to fast-forward a DataPipe. Usage of other fast-forwarding
+        methods (custom ones if necessary) are recommended.
+
+    Args:
+        datapipe: IterDataPipe to be fast-forwarded
+        n_iterations: number of iterations to fast-forward
+        rng: ``Optional[torch.Generator]``. If not ``None``, this RNG will be used for shuffling. The generator
+            should be in its `initial` state as it was first passed into ``DataLoader`` or ``ReadingService``.
+    """
+    # This operation can only take place if `datapipe` has been recently restored (usually through `__setstate__`)
+    # and an iterator has not started yet.
+    if not datapipe._restored:
+        return
```

Comment:
I am still confused here. `_restored` indicates the DataPipe is restored from snapshot.
Then, after the DataPipe is restored, we still want to do fast forwarding here. Why not we directly fast-forward when restoring from snapshot?

Reply:
Perhaps we should rename `_restored`, currently two built-in things can set `_restored = True`:
1. `__setstate__` (usually called by deserialization)
2. The function that we are adding in this PR

Reply:
I think perhaps we need 3 states:
1. `Deserialized` - allows you to restore snapshot and create an iterator without reset
2. `Restored` - cannot restore again, allows you to create an iterator without reset
3. `Iterating`? - cannot restore, will reset if you create a new iterator

Reply:
For the state 3, we should allow restore during iterating. I think normally we would do:
```py
it = iter(dl)
for _ in range(10):
    next(it)

state = dl.state_dict()

dl.restore(state  # Restore during iteration?
it = iter(dl)
```

Reply:
I think it is feasible to support this, but how common is this and why is this done? 
Note that the original `iterator` will be invalidated during the restoration.

@VitalyFedyunin Any thoughts?

Reply:
Note that I updated the PR with the 3 states as discussed above. Thanks for the feedback!